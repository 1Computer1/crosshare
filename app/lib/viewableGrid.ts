import {
  GridBase, EntryBase,
  posForIndex, cellIndex, valAt, entryAtPosition, entriesFromCells
} from './gridBase';
import { ClueT, Position, Direction, PosAndDir, BLOCK } from './types';
import { Symmetry } from '../reducers/reducer';

export interface ViewableEntry extends EntryBase {
  labelNumber: number,
}

export interface CluedEntry extends ViewableEntry {
  clue: string,
}

export interface ViewableGrid<Entry extends ViewableEntry> extends GridBase<Entry> {
  sortedEntries: Array<number>;
  cellLabels: Map<number, number>,
  allowBlockEditing: boolean,
  highlighted: Set<number>,
  highlight: 'circle' | 'shade',
  mapper(entry: ViewableEntry): Entry,
}

export interface CluedGrid extends ViewableGrid<CluedEntry> {
  clues: Array<ClueT>,
}

export function getSortedEntries<Entry extends EntryBase>(entries: Array<Entry>) {
  return [...entries].sort((a, b) => {
    if (a.direction !== b.direction) {
      return a.direction - b.direction;
    }
    return a.index - b.index;
  }).map(e => e.index);
}

export function moveLeft<Entry extends ViewableEntry>(grid: ViewableGrid<Entry>, active: Position): Position {
  let x = active.col;
  while (x >= 0) {
    x -= 1;
    if (x >= 0 && (grid.allowBlockEditing || valAt(grid, { ...active, col: x }) !== BLOCK)) {
      break;
    }
  }
  if (x < 0) {
    return active;
  }
  return { ...active, col: x };
}

export function moveRight<Entry extends ViewableEntry>(grid: ViewableGrid<Entry>, active: Position): Position {
  let x = active.col;
  while (x < grid.width) {
    x += 1;
    if (x < grid.width && (grid.allowBlockEditing || valAt(grid, { ...active, col: x }) !== BLOCK)) {
      break;
    }
  }
  if (x >= grid.width) {
    return active;
  }
  return { ...active, col: x };
}

export function moveUp<Entry extends ViewableEntry>(grid: ViewableGrid<Entry>, active: Position): Position {
  let y = active.row;
  while (y >= 0) {
    y -= 1;
    if (y >= 0 && (grid.allowBlockEditing || valAt(grid, { ...active, row: y }) !== BLOCK)) {
      break;
    }
  }
  if (y < 0) {
    return active;
  }
  return { ...active, row: y };
}

export function moveDown<Entry extends ViewableEntry>(grid: ViewableGrid<Entry>, active: Position): Position {
  let y = active.row;
  while (y < grid.height) {
    y += 1;
    if (y < grid.height && (grid.allowBlockEditing || valAt(grid, { ...active, row: y }) !== BLOCK)) {
      break;
    }
  }
  if (y >= grid.height) {
    return active;
  }
  return { ...active, row: y };
}

export function retreatPosition<Entry extends ViewableEntry>(grid: ViewableGrid<Entry>, pos: PosAndDir): PosAndDir {
  const [entry, index] = entryAtPosition(grid, pos);
  if (entry !== null && index > 0) {
    return { ...entry.cells[index - 1], dir: pos.dir };
  }
  if (grid.allowBlockEditing) {
    const xincr = (pos.dir === Direction.Across) ? -1 : 0;
    const yincr = (pos.dir === Direction.Down) ? -1 : 0;
    if ((pos.row + yincr >= 0) && (pos.col + xincr >= 0)) {
      return { row: pos.row + yincr, col: pos.col + xincr, dir: pos.dir };
    }
  }
  return pos;
}

export function nextCell<Entry extends ViewableEntry>(grid: ViewableGrid<Entry>, pos: PosAndDir): PosAndDir {
  const [entry, index] = entryAtPosition(grid, pos);
  if (entry !== null && index < entry.cells.length - 1) {
    return { ...entry.cells[index + 1], dir: pos.dir };
  }
  if (grid.allowBlockEditing) {
    const xincr = (pos.dir === Direction.Across) ? 1 : 0;
    const yincr = (pos.dir === Direction.Down) ? 1 : 0;
    if ((pos.row + yincr < grid.height) && (pos.col + xincr < grid.width)) {
      return { row: pos.row + yincr, col: pos.col + xincr, dir: pos.dir };
    }
  }
  return pos;
}

export function nextNonBlock<Entry extends ViewableEntry>(grid: ViewableGrid<Entry>, pos: Position) {
  const index = cellIndex(grid, pos);
  for (let offset = 0; offset < grid.cells.length; offset += 1) {
    if (grid.cells[(index + offset) % grid.cells.length] !== BLOCK) {
      return posForIndex(grid, index + offset);
    }
  }
  return pos;
}

export function advancePosition<Entry extends ViewableEntry>(grid: ViewableGrid<Entry>, pos: PosAndDir, wrongCells: Set<number>): PosAndDir {
  const [entry, index] = entryAtPosition(grid, pos);
  if (!entry) {
    return pos;
  }
  for (let offset = 0; offset < entry.cells.length; offset += 1) {
    const cell = entry.cells[(index + offset + 1) % entry.cells.length];
    if (valAt(grid, cell) === ' ' || wrongCells.has(cellIndex(grid, cell))) {
      return { ...cell, dir: pos.dir };
    }
  }
  if (index === entry.cells.length - 1) {
    return moveToNextEntry(grid, pos);
  }
  return { ...entry.cells[index + 1], dir: pos.dir };
}

export function moveToPrevEntry<Entry extends ViewableEntry>(grid: ViewableGrid<Entry>, pos: PosAndDir): PosAndDir {
  return moveToNextEntry(grid, pos, true);
}

export function moveToNextEntry<Entry extends ViewableEntry>(grid: ViewableGrid<Entry>, pos: PosAndDir, reverse = false): PosAndDir {
  const [currentEntry,] = entryAtPosition(grid, pos);
  if (!currentEntry) {
    const xincr = (pos.dir === Direction.Across) ? 1 : 0;
    const yincr = (pos.dir === Direction.Down) ? 1 : 0;
    let iincr = xincr + yincr * grid.width;
    if (reverse) {
      iincr *= -1;
    }
    return { ...posForIndex(grid, (cellIndex(grid, pos) + iincr) % (grid.width * grid.height)), dir: pos.dir };
  }

  // Find position in the sorted array of entries
  let i = 0;
  for (; i < grid.sortedEntries.length; i += 1) {
    if (currentEntry.index === grid.sortedEntries[i]) {
      break;
    }
  }

  // Now find the next entry in the sorted array that isn't complete
  for (let offset = 0; offset < grid.sortedEntries.length; offset += 1) {
    let index = (i + offset + 1) % grid.sortedEntries.length;
    if (reverse) {
      index = (i + grid.sortedEntries.length - offset - 1) % grid.sortedEntries.length;
    }
    const tryEntry = grid.entries[grid.sortedEntries[index]];
    if (tryEntry.completedWord === null) {
      for (const cell of tryEntry.cells) {
        if (valAt(grid, cell) === ' ') {
          return { ...cell, dir: tryEntry.direction };
        }
      }
    }
  }

  // Now just return the start of the next/prev entry
  let index = (i + 1) % grid.sortedEntries.length;
  if (reverse) {
    index = (i - 1) % grid.sortedEntries.length;
  }
  const nextEntry = grid.entries[grid.sortedEntries[index]];
  return { ...nextEntry.cells[0], dir: nextEntry.direction };
}

export function gridWithNewChar<Entry extends ViewableEntry,
  Grid extends ViewableGrid<Entry>>(grid: Grid, pos: Position, char: string, sym: Symmetry): Grid {
  const index = pos.row * grid.width + pos.col;
  const cells = [...grid.cells];
  if (valAt(grid, pos) === BLOCK) {
    if (!grid.allowBlockEditing) {
      return grid;
    }

    const flippedCell = flipped(grid, pos, sym);
    if (flippedCell !== null && cells[flippedCell] === BLOCK) {
      cells[flippedCell] = ' ';
    }
  }
  cells[index] = char;
  return fromCells({ ...grid, cells });
}

function flipped<Entry extends ViewableEntry,
  Grid extends ViewableGrid<Entry>>(grid: Grid, pos: Position, sym: Symmetry): number | null {
  switch (sym) {
  case Symmetry.None:
    return null;
  case Symmetry.Rotational:
    return (grid.height - pos.row - 1) * grid.width + (grid.width - pos.col - 1);
  case Symmetry.Horizontal:
    return (grid.height - pos.row - 1) * grid.width + pos.col;
  case Symmetry.Vertical:
    return pos.row * grid.width + (grid.width - pos.col - 1);
  case Symmetry.DiagonalNESW:
    return (grid.height - pos.col - 1) * grid.width + (grid.width - pos.row - 1);
  case Symmetry.DiagonalNWSE:
    return pos.col * grid.width + pos.row;
  }
}

export function gridWithBlockToggled<Entry extends ViewableEntry,
  Grid extends ViewableGrid<Entry>>(grid: Grid, pos: Position, sym: Symmetry): Grid {
  let char = BLOCK;
  if (valAt(grid, pos) === BLOCK) {
    char = ' ';
  }
  const index = pos.row * grid.width + pos.col;
  const cells = [...grid.cells];
  cells[index] = char;

  const flippedCell = flipped(grid, pos, sym);
  if (flippedCell !== null) {
    cells[flippedCell] = char;
  }

  return fromCells({ ...grid, cells });
}

function cluesByDirection(rawClues: Array<ClueT>) {
  const clues = [new Map<number, string>(), new Map<number, string>()];
  for (const clue of rawClues) {
    clues[clue.dir].set(clue.num, clue.clue);
  }
  return clues;
}

export function addClues<Entry extends ViewableEntry,
  Grid extends ViewableGrid<Entry>>(
  grid: Grid, rawClues: Array<ClueT>): CluedGrid {
  const clues = cluesByDirection(rawClues);

  function mapper(e: Entry): CluedEntry {
    const clue = clues[e.direction].get(e.labelNumber);
    if (!clue) {
      throw new Error('Can\'t find clue for ' + e.labelNumber + ' ' + e.direction);
    }
    return { ...e, clue };
  }

  const entries: Array<CluedEntry> = [];
  for (const entry of grid.entries) {
    entries.push(mapper(entry));
  }
  return { ...grid, mapper: (e) => mapper(grid.mapper(e)), clues: rawClues, entries };
}

export function fromCells<Entry extends ViewableEntry,
  Grid extends ViewableGrid<Entry>>(
  input: Omit<Grid, 'entries' | 'entriesByCell' | 'sortedEntries' | 'cellLabels'>
): Grid {

  const [baseEntries, entriesByCell] = entriesFromCells(input.width, input.height, input.cells);

  const cellLabels = new Map<number, number>();
  let currentCellLabel = 1;
  const entries: Array<Entry> = [];
  for (const baseEntry of baseEntries) {
    const startPos = baseEntry.cells[0];
    const i = startPos.row * input.width + startPos.col;
    const entryLabel = cellLabels.get(i) || currentCellLabel;
    if (!cellLabels.has(i)) {
      cellLabels.set(i, currentCellLabel);
      currentCellLabel += 1;
    }
    entries.push(input.mapper({
      ...baseEntry,
      labelNumber: entryLabel,
    }));
  }
  return {
    ...input,
    sortedEntries: getSortedEntries(entries),
    entriesByCell,
    entries,
    cellLabels,
  } as Grid; // TODO remove assertion after this is fixed: https://github.com/microsoft/TypeScript/issues/28884#issuecomment-448356158
}
