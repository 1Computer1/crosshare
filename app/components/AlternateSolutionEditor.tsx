import useEventListener from '@use-it/event-listener';
import { useRef, useCallback, useReducer, useMemo } from 'react';
import { FaSave, FaWindowClose, FaEllipsisH, FaVolumeUp, FaVolumeMute, FaKeyboard } from 'react-icons/fa';
import { usePersistedBoolean } from '../lib/hooks';
import { Direction, fromKeyboardEvent, fromKeyString, KeyK } from '../lib/types';
import { fromCells } from '../lib/viewableGrid';
import { Rebus, EscapeKey } from './Icons';
import { Keyboard } from './Keyboard';
import { TopBar, TopBarLink, TopBarDropDown, TopBarDropDownLink } from './TopBar';
import { gridInterfaceReducer, KeypressAction, PasteAction } from '../reducers/reducer';
import { Square } from './Square';
import { GridView } from './Grid';
import { isSome } from 'fp-ts/lib/Option';

export function AlternateSolutionEditor(props: { grid: string[], width: number, height: number, highlighted: Set<number>, highlight: 'circle' | 'shade', cancel: () => void, save: (alt: Record<number, string>) => Promise<void> }) {
  const initialGrid = fromCells({
    mapper: (e) => e,
    width: props.width,
    height: props.height,
    cells: props.grid,
    allowBlockEditing: false,
    highlighted: props.highlighted,
    highlight: props.highlight,
  });

  const [state, dispatch] = useReducer(gridInterfaceReducer, {
    type: 'alternate-editor',
    active: { col: 0, row: 0, dir: Direction.Across },
    grid: initialGrid,
    wasEntryClick: false,
    showExtraKeyLayout: false,
    isEnteringRebus: false,
    rebusValue: '',
    downsOnly: false,
    isEditable: () => true,
  });

  const [muted, setMuted] = usePersistedBoolean('muted', false);
  const [toggleKeyboard, setToggleKeyboard] = usePersistedBoolean(
    'keyboard',
    false
  );

  const gridRef = useRef<HTMLDivElement | null>(null);

  const focusGrid = useCallback(() => {
    if (gridRef.current) {
      gridRef.current.focus();
    }
  }, []);

  const physicalKeyboardHandler = useCallback(
    (e: KeyboardEvent) => {
      const tagName = (e.target as HTMLElement)?.tagName?.toLowerCase();
      if (tagName === 'textarea' || tagName === 'input') {
        return;
      }
      const mkey = fromKeyboardEvent(e);
      if (isSome(mkey)) {
        const kpa: KeypressAction = { type: 'KEYPRESS', key: mkey.value };
        dispatch(kpa);
        e.preventDefault();
      }
    },
    [dispatch]
  );
  useEventListener(
    'keydown',
    physicalKeyboardHandler,
    gridRef.current || undefined
  );

  const pasteHandler = useCallback((e: ClipboardEvent) => {
    const tagName = (e.target as HTMLElement)?.tagName?.toLowerCase();
    if (tagName === 'textarea' || tagName === 'input') {
      return;
    }
    const pa: PasteAction = {
      type: 'PASTE',
      content: e.clipboardData?.getData('Text') || ''
    };
    dispatch(pa);
    e.preventDefault();
  }, [dispatch]);
  useEventListener(
    'paste',
    pasteHandler
  );

  const keyboardHandler = useCallback(
    (key: string) => {
      const mkey = fromKeyString(key);
      if (isSome(mkey)) {
        const kpa: KeypressAction = { type: 'KEYPRESS', key: mkey.value };
        dispatch(kpa);
      }
    },
    [dispatch]
  );

  const topBarChildren = useMemo(() => {
    return (
      <>
        <TopBarLink
          icon={<FaSave />}
          text="Add Alternate"
          onClick={async () => {
            const alt: Record<number, string> = {};
            let hadAny = false;
            for (const [idx, cellValue] of state.grid.cells.entries()) {
              const defaultCellValue = initialGrid.cells[idx];
              if (cellValue.trim() && cellValue.trim() != defaultCellValue?.trim()) {
                hadAny = true;
                alt[idx] = cellValue;
              }
            }
            if (!hadAny) {
              props.cancel();
              return;
            }
            return props.save(alt).then(() => props.cancel());
          }}
        />
        <TopBarLink
          icon={<FaWindowClose />}
          text="Cancel"
          onClick={props.cancel}
        />
        <TopBarDropDown onClose={focusGrid} icon={<FaEllipsisH />} text="More">
          {() => (
            <>
              <TopBarDropDownLink
                icon={<Rebus />}
                text="Enter Rebus"
                shortcutHint={<EscapeKey />}
                onClick={() => {
                  const a: KeypressAction = {
                    type: 'KEYPRESS',
                    key: { k: KeyK.Escape },
                  };
                  dispatch(a);
                }}
              />
              {muted ? (
                <TopBarDropDownLink
                  icon={<FaVolumeUp />}
                  text="Unmute"
                  onClick={() => setMuted(false)}
                />
              ) : (
                <TopBarDropDownLink
                  icon={<FaVolumeMute />}
                  text="Mute"
                  onClick={() => setMuted(true)}
                />
              )}
              <TopBarDropDownLink
                icon={<FaKeyboard />}
                text="Toggle Keyboard"
                onClick={() => setToggleKeyboard(!toggleKeyboard)}
              />
            </>
          )}
        </TopBarDropDown>
      </>
    );
  }, [focusGrid, initialGrid.cells, muted, props, setMuted, setToggleKeyboard, state.grid.cells, toggleKeyboard]);

  const parentRef = useRef<HTMLDivElement | null>(null);

  return (
    <>
      <div
        css={{
          display: 'flex',
          flexDirection: 'column',
          height: '100%',
        }}
      >
        <div css={{ flex: 'none' }}>
          <TopBar>{topBarChildren}</TopBar>
        </div>
        <div
          css={{ flex: '1 1 auto', overflow: 'scroll', position: 'relative' }}
        >
          <div
            // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
            tabIndex={0}
            ref={(instance) => {
              parentRef.current = instance;
            }}
            css={{
              outline: 'none',
              display: 'flex',
              flex: '1 1 auto',
              flexDirection: 'column',
              alignItems: 'center',
              height: '100%',
              width: '100%',
              position: 'absolute',
              flexWrap: 'nowrap',
            }}
          >
            <Square
              noColumns={true}
              waitToResize={false}
              parentRef={parentRef}
              aspectRatio={props.width / props.height}
              contents={(width: number, _height: number) => {
                return (
                  <GridView
                    isEnteringRebus={state.isEnteringRebus}
                    rebusValue={state.rebusValue}
                    squareWidth={width}
                    grid={state.grid}
                    defaultGrid={initialGrid}
                    active={state.active}
                    dispatch={dispatch}
                    allowBlockEditing={false}
                    autofill={[]}
                  />
                );
              }}
            />
          </div>
        </div>
        <div css={{ flex: 'none', width: '100%' }}>
          <Keyboard
            toggleKeyboard={toggleKeyboard}
            keyboardHandler={keyboardHandler}
            muted={muted}
            showExtraKeyLayout={state.showExtraKeyLayout}
            includeBlockKey={false}
          />
        </div>
      </div>
    </>
  );
}