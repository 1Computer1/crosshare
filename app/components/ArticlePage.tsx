import Head from 'next/head';
import { ArticleT } from '../lib/article';
import { ArticlePageProps } from '../lib/serverOnly';
import { HUGE_AND_UP, MAX_WIDTH } from '../lib/style';
import { ErrorPage } from './ErrorPage';
import { Link } from './Link';
import { Markdown } from './Markdown';
import { DefaultTopBar } from './TopBar';

export function ArticlePage(props: ArticlePageProps) {
  if ('error' in props) {
    return (
      <ErrorPage title="Article Not Found">
        <p>
          We&apos;re sorry, we couldn&apos;t find the article you requested.
        </p>
        <p>{props.error}</p>
        <p>
          Try the <Link href="/">homepage</Link>.
        </p>
      </ErrorPage>
    );
  }
  return <Article key={props.s} {...props} />;
}

function Article(props: ArticleT) {
  return (
    <>
      <Head>
        <title>{props.t}</title>
        <meta key="og:title" property="og:title" content={props.t} />
        <link
          rel="canonical"
          href={'https://crosshare.org/article/' + props.s}
        />
      </Head>
      <DefaultTopBar />
      <Markdown
        text={props.c}
        css={{
          margin: '2em 1em',
          [HUGE_AND_UP]: {
            maxWidth: MAX_WIDTH,
            margin: '2em auto',
          },
        }}
      />
    </>
  );
}
