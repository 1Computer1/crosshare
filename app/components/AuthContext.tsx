import { ReactNode, createContext, useContext } from 'react';
import { ErrorPage } from '../components/ErrorPage';
import type firebase from 'firebase/app';
import { GoogleLinkButton, GoogleSignInButton } from './GoogleButtons';
import { TopBar } from './TopBar';
import { ConstructorPageT } from '../lib/constructorPage';
import { NotificationT } from '../lib/notifications';
import { AccountPrefsT } from '../lib/prefs';
import { WithConditionalCSSProp } from '@emotion/react/types/jsx-namespace';

export interface AuthProps {
  isAdmin: boolean;
  user: firebase.User;
  constructorPage?: ConstructorPageT;
  prefs?: AccountPrefsT;
}

export interface AuthPropsOptional {
  isAdmin: boolean;
  user?: firebase.User;
  constructorPage?: ConstructorPageT;
  prefs?: AccountPrefsT;
}

export function renderLoginButtonIfNeeded({
  user,
  loading,
  error,
}: AuthContextValue): React.ReactNode | null {
  if (loading) {
    return <div></div>;
  }
  if (error) {
    return <div>Error loading user: {error}</div>;
  }
  if (!user) {
    return <GoogleSignInButton />;
  }
  if (user.isAnonymous) {
    return <GoogleLinkButton user={user} />;
  }
  return null;
}

function renderLoginIfNeeded({
  user,
  loading,
  error,
}: AuthContextValue): React.ReactNode | null {
  if (loading) {
    return <div></div>;
  }
  if (error) {
    return <div>Error loading user: {error}</div>;
  }
  if (!user) {
    return (
      <>
        <TopBar />
        <div css={{ margin: '1em' }}>
          <p>
            Please sign-in with your Google account to continue. We use your
            account to keep track of the puzzles you&apos;ve played and your
            solve streaks.
          </p>
          <GoogleSignInButton />
        </div>
      </>
    );
  }
  if (user.isAnonymous) {
    return (
      <>
        <TopBar />
        <div css={{ margin: '1em' }}>
          <p>
            Please sign-in with your Google account to continue. We use your
            account to keep track of the puzzles you&apos;ve played and your
            solve streaks.
          </p>
          <GoogleLinkButton user={user} />
        </div>
      </>
    );
  }
  return null;
}

/* Ensure we have a non-anonymous user, upgrading an anonymous user if we have one. */
export function requiresAuth<T>(WrappedComponent: React.ComponentType<T>) {
  return (props: T & WithConditionalCSSProp<T>): ReactNode => {
    const ctx = useContext(AuthContext);
    const login = renderLoginIfNeeded(ctx);
    if (login) {
      return login;
    }
    return (
      <WrappedComponent
        {...props}
        isAdmin={ctx.isAdmin}
        user={ctx.user}
        constructorPage={ctx.constructorPage}
        prefs={ctx.prefs}
      />
    );
  };
}

/* Ensure we have an admin user, upgrading an anonymous user if we have one. */
export function requiresAdmin<T>(WrappedComponent: React.ComponentType<T>) {
  return (props: T & WithConditionalCSSProp<T>): ReactNode => {
    const ctx = useContext(AuthContext);
    const login = renderLoginIfNeeded(ctx);
    if (login) {
      return login;
    }
    if (!ctx.isAdmin) {
      return (
        <ErrorPage title="Not Allowed">
          <p>You do not have permission to view this page</p>
        </ErrorPage>
      );
    }
    return (
      <WrappedComponent
        {...props}
        isAdmin={true}
        user={ctx.user}
        constructorPage={ctx.constructorPage}
        prefs={ctx.prefs}
      />
    );
  };
}

interface AuthContextValue {
  user?: firebase.User;
  notifications?: Array<NotificationT>;
  isAdmin: boolean;
  loading: boolean;
  error?: string;
  constructorPage?: ConstructorPageT;
  prefs?: AccountPrefsT;
}
export const AuthContext = createContext({
  user: undefined,
  loading: false,
  error: 'using default context',
  constructorPage: undefined,
  prefs: undefined,
} as AuthContextValue);
