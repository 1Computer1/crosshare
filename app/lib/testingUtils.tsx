import { ReactElement, ComponentType } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { matchers, createSerializer } from 'jest-emotion';

import { AuthContext } from '../components/AuthContext';

import type firebaseTypes from 'firebase';

// Add the custom matchers provided by 'jest-emotion'
expect.extend(matchers);
expect.addSnapshotSerializer(createSerializer());

export const anonymousUser = {
  uid: 'anonymous-user-id',
  isAnonymous: true
} as firebaseTypes.User;

const WithAllProviders: ComponentType = ({ children }) => {
  return (
    <AuthContext.Provider value={{ user: anonymousUser, isAdmin: false, loadingUser: false, error: undefined }}>
      {children}
    </AuthContext.Provider>
  )
}

function wrappedRender(
  ui: ReactElement,
  options?: RenderOptions
): RenderResult {

  return render(ui, {
    wrapper: WithAllProviders,
    ...options,
  });
}
// re-export everything
export * from '@testing-library/react';

// override render method
export { wrappedRender as render };
