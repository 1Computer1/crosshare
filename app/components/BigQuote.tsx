import { SECONDARY } from '../lib/style';
import { ReactNode } from 'react';

export function BigQuote(props: { quote: string, attribution: ReactNode }) {
  return <div css={{
    padding: '2em',
    backgroundColor: SECONDARY,
    textAlign: 'center',
    color: 'var(--text)',
  }}>
    <div css={{
      maxWidth: 900,
      margin: 'auto',
      '&:before': {
        fontFamily: 'Georgia, serif',
        content: 'open-quote',
        fontSize: '6em',
        lineHeight: '0.1em',
        marginRight: '0.25em',
        verticalAlign: '-0.4em',
      }
    }}>
      <span css={{ fontFamily: 'Georgia, serif', fontSize: 30, fontStyle: 'italic' }}>{props.quote}</span>
      <br />
      <span css={{ fontSize: 20 }}>&mdash; {props.attribution}</span>
    </div>
  </div>;
}
