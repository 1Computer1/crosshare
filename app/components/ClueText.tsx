import { ReactNode, Fragment } from 'react';

import { CluedEntry, RefPosition } from '../lib/viewableGrid';
import { GridBase, valAt, EntryBase } from '../lib/gridBase';

import { ToolTipText } from './ToolTipText';
import { getClueText } from '../lib/types';

interface ClueTextProps {
  entryIndex: number;
  allEntries: Array<CluedEntry>;
  refPositions: Array<Array<RefPosition>>;
  grid: GridBase<EntryBase>;
}
export const ClueText = (props: ClueTextProps) => {
  const entry = props.allEntries[props.entryIndex];
  if (!entry) {
    throw new Error('oob');
  }
  const text = getClueText(entry);
  let offset = 0;
  const parts: Array<ReactNode> = [];
  let i = 0;
  for (const [refIndex, start, end] of props.refPositions[props.entryIndex] ||
    []) {
    if (offset < start) {
      parts.push(<Fragment key={i++}>{text.slice(offset, start)}</Fragment>);
    }
    const e = props.allEntries[refIndex];
    if (!e) {
      throw new Error('oob');
    }
    parts.push(
      <ToolTipText
        key={i++}
        text={text.slice(start, end)}
        tooltip={
          <>
            {getClueText(e)}
            <b
              css={{
                marginLeft: '0.5em',
                whiteSpace: 'nowrap',
              }}
            >
              [{e.cells.map((a) => valAt(props.grid, a).trim() || '-')}]
            </b>
          </>
        }
      />
    );
    offset = end;
  }
  if (offset < text.length) {
    parts.push(<Fragment key={i++}>{text.slice(offset)}</Fragment>);
  }
  return <div>{parts}</div>;
};
