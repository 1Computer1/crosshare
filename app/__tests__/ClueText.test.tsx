import { act, render } from '../lib/testingUtils';
import { addClues, CluedGrid, fromCells, getRefs } from '../lib/viewableGrid';
import { ClueText } from '../components/ClueText';

test('Highlighting for different types of clues', async () => {
  const answers = ['U', 'P', 'S', 'O'];
  const grid = fromCells({
    width: 2,
    height: 2,
    cells: answers,
    allowBlockEditing: false,
    highlighted: new Set<number>(),
    vBars: new Set<number>(),
    hBars: new Set<number>(),
    highlight: 'circle',
    mapper: (x) => x,
  });

  const cluedGrid: CluedGrid = addClues(grid, [
    { num: 1, dir: 0, clue: 'Not down 5-acrosses', explanation: null },
    { num: 3, dir: 0, clue: '2-down Then...', explanation: null },
    { num: 1, dir: 1, clue: '1- and 3- acrosses You and I', explanation: null },
    { num: 2, dir: 1, clue: '3A Post office abbr.', explanation: null },
  ]);

  const [, refPositions] = getRefs(cluedGrid);

  console.log(refPositions);
  for (let i = 0; i < 4; i += 1) {
    const { container } = render(
      <div>
        <ClueText
          grid={cluedGrid}
          allEntries={cluedGrid.entries}
          entryIndex={i}
          refPositions={refPositions}
          downsOnly={false}
        />
      </div>,
      {}
    );
    expect(container).toMatchSnapshot();
    await act(async () => Promise.resolve()); // Popper update() - https://github.com/popperjs/react-popper/issues/350
  }
});
