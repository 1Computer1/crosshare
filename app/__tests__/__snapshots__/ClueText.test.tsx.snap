// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlighting for different types of clues 1`] = `
<div>
  <div>
    Not down 5-acrosses
  </div>
</div>
`;

exports[`Highlighting for different types of clues 2`] = `
.emotion-0 {
  border-bottom: 1px dotted;
  white-space: nowrap;
}

.emotion-1 {
  z-index: 100000;
  border-radius: 5px;
  background-color: var(--black);
  color: var(--white);
  text-align: center;
  max-width: 30em;
  padding: 10px;
  visibility: hidden;
}

.emotion-1[data-popper-reference-hidden=true] {
  visibility: hidden;
}

.emotion-2 {
  margin-left: 0.5em;
  white-space: nowrap;
}

.emotion-3 {
  position: absolute;
  width: 10px;
  height: 10px;
}

[data-popper-placement^="bottom"] .emotion-3 {
  top: -5px;
}

[data-popper-placement^="top"] .emotion-3 {
  bottom: -5px;
}

.emotion-3::after {
  content: " ";
  position: absolute;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 10px;
  height: 10px;
  background-color: var(--black);
}

<div>
  <div>
    <span
      class="emotion-0"
    >
      1
    </span>
    <div
      class="emotion-1"
      style="position: fixed; left: 0px; top: 0px;"
    >
      Not down 5-acrosses
      <b
        class="emotion-2"
      >
        [
        U
        P
        ]
      </b>
      <div
        class="emotion-3"
        style="position: absolute;"
      />
    </div>
    - and 
    <span
      class="emotion-0"
    >
      3- acrosses
    </span>
    <div
      class="emotion-1"
      style="position: fixed; left: 0px; top: 0px;"
    >
      2-down Then...
      <b
        class="emotion-2"
      >
        [
        S
        O
        ]
      </b>
      <div
        class="emotion-3"
        style="position: absolute;"
      />
    </div>
     You and I
  </div>
</div>
`;

exports[`Highlighting for different types of clues 3`] = `
.emotion-0 {
  border-bottom: 1px dotted;
  white-space: nowrap;
}

.emotion-1 {
  z-index: 100000;
  border-radius: 5px;
  background-color: var(--black);
  color: var(--white);
  text-align: center;
  max-width: 30em;
  padding: 10px;
  visibility: hidden;
}

.emotion-1[data-popper-reference-hidden=true] {
  visibility: hidden;
}

.emotion-2 {
  margin-left: 0.5em;
  white-space: nowrap;
}

.emotion-3 {
  position: absolute;
  width: 10px;
  height: 10px;
}

[data-popper-placement^="bottom"] .emotion-3 {
  top: -5px;
}

[data-popper-placement^="top"] .emotion-3 {
  bottom: -5px;
}

.emotion-3::after {
  content: " ";
  position: absolute;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 10px;
  height: 10px;
  background-color: var(--black);
}

<div>
  <div>
    <span
      class="emotion-0"
    >
      3A
    </span>
    <div
      class="emotion-1"
      style="position: fixed; left: 0px; top: 0px;"
    >
      2-down Then...
      <b
        class="emotion-2"
      >
        [
        S
        O
        ]
      </b>
      <div
        class="emotion-3"
        style="position: absolute;"
      />
    </div>
     Post office abbr.
  </div>
</div>
`;

exports[`Highlighting for different types of clues 4`] = `
.emotion-0 {
  border-bottom: 1px dotted;
  white-space: nowrap;
}

.emotion-1 {
  z-index: 100000;
  border-radius: 5px;
  background-color: var(--black);
  color: var(--white);
  text-align: center;
  max-width: 30em;
  padding: 10px;
  visibility: hidden;
}

.emotion-1[data-popper-reference-hidden=true] {
  visibility: hidden;
}

.emotion-2 {
  margin-left: 0.5em;
  white-space: nowrap;
}

.emotion-3 {
  position: absolute;
  width: 10px;
  height: 10px;
}

[data-popper-placement^="bottom"] .emotion-3 {
  top: -5px;
}

[data-popper-placement^="top"] .emotion-3 {
  bottom: -5px;
}

.emotion-3::after {
  content: " ";
  position: absolute;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 10px;
  height: 10px;
  background-color: var(--black);
}

<div>
  <div>
    <span
      class="emotion-0"
    >
      2-down
    </span>
    <div
      class="emotion-1"
      style="position: fixed; left: 0px; top: 0px;"
    >
      3A Post office abbr.
      <b
        class="emotion-2"
      >
        [
        P
        O
        ]
      </b>
      <div
        class="emotion-3"
        style="position: absolute;"
      />
    </div>
     Then...
  </div>
</div>
`;
